examples/cpp/helloworld/greeter_async_bidi_client.cc

```diff
@@ -0,0 +1,213 @@
+    HelloRequest request;
+    request.set_name(user);
+
+    // This is important: You can have at most one write or at most one read
+    // at any given time. The throttling is performed by gRPC completion
+    // queue. If you queue more than one write/read, the stream will crash.
+    // Because this stream is bidirectional, you *can* have a single read
+    // and a single write request queued for the same stream. Writes and reads
+    // are independent of each other in terms of ordering/delivery.
+    std::cout << " ** Sending request: " << user << std::endl;
+    stream_->Write(request, reinterpret_cast<void*>(Type::WRITE));
+    return true;
+  }
+
+  ~AsyncBidiGreeterClient() {
+    std::cout << "Shutting down client...." << std::endl;
+    grpc::Status status;
+    cq_.Shutdown();
+    grpc_thread_->join();
+  }
+
+ private:
+  void AsyncHelloRequestNextMessage() {
+    std::cout << " ** Got response: " << response_.message() << std::endl;
+
+    // The tag is the link between our thread (main thread) and the completion
+    // queue thread. The tag allows the completion queue to fan off
+    // notification handlers for the specified read/write requests as they
+    // are being processed by gRPC.
+    stream_->Read(&response_, reinterpret_cast<void*>(Type::READ));
+  }
+
+  // Runs a gRPC completion-queue processing thread. Checks for 'Next' tag
+  // and processes them until there are no more (or when the completion queue
+  // is shutdown).
+  void GrpcThread() {
+    while (true) {
+      void* got_tag;
+      bool ok = false;
+      // Block until the next result is available in the completion queue "cq".
+      // The return value of Next should always be checked. This return value
+      // tells us whether there is any kind of event or the cq_ is shutting
+      // down.
+      if (!cq_.Next(&got_tag, &ok)) {
+        std::cerr << "Client stream closed. Quitting" << std::endl;
+        break;
+      }
+
+      // It's important to process all tags even if the ok is false. One might
+      // want to deallocate memory that has be reinterpret_cast'ed to void*
+      // when the tag got initialized. For our example, we cast an int to a
+      // void*, so we don't have extra memory management to take care of.
+      if (ok) {
+        std::cout << std::endl
+                  << "**** Processing completion queue tag " << got_tag
+                  << std::endl;
+        switch (static_cast<Type>(reinterpret_cast<long>(got_tag))) {
+          case Type::READ:
+            std::cout << "Read a new message." << std::endl;
+            break;
+          case Type::WRITE:
+            std::cout << "Sending message (async)." << std::endl;
+            AsyncHelloRequestNextMessage();
+            break;
+          case Type::CONNECT:
+            std::cout << "Server connected." << std::endl;
+            break;
+          case Type::WRITES_DONE:
+            std::cout << "Server disconnecting." << std::endl;
+            break;
```

Comment:
So there  might be an arbitrary gap between Writes_done completion and actual server disconnection depending on what the server wants to do when its read fails; it might send 100 more messages at that point in general.